### PR TITLE
Qmi5 add type annotations

### PR DESCRIPTION
**Pull request by qraa**

### CODE REVIEW

File: qmi/tools/proc.py [hunk 1]
```diff
@@ -893,10 +904,7 @@ def main() -> int:
 
     :return: 0 in case of success; exit status in case of error.
     """
-
-    parser = argparse.ArgumentParser()
-    parser.format_help = lambda: __doc__ + "\n"
-    parser.format_usage = lambda: __doc__ + "\n"
+    parser = argparse.ArgumentParser(description=main.__doc__)
```
Comment (jorisvr) [inline qmi/tools/proc.py]: This is not the same. The `description` is meant to be a single paragraph describing the program; and command-line options to be documented in a different way.

I prefer to have a complete usage instruction in the docstring at the top of the main program. This does not match the philosophy of ArgumentParser, therefore I work around by overriding the two formatting methods. We can have a discussion about the best way to handle argument parsing and help texts in QMI. But I don't think this change is ok in the scope of type annotations.

File: tests/core/test_basic.py [hunk 2]
```diff
@@ -15,8 +15,8 @@ class TestQmiBasic(unittest.TestCase):
     def test_start_stop(self):
 
         # Check qmi.context() returns None when there is no context.
-        ctx = qmi.context()
-        self.assertIsNone(ctx)
+        with self.assertRaises(qmi.core.exceptions.QMI_NoActiveContextException):
+            ctx = qmi.context()
```
Comment (jorisvr) [inline tests/core/test_basic.py]: Change comment to reflect the new semantics.

File: tests/tools/test_proc.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import unittest
+
+from qmi.tools.proc import is_local_host
+
+
+class ProcTestCase(unittest.TestCase):
+    
+    def test_is_local_host(self):
+        self.assertFalse(is_local_host("0.0.0.0"))
```
Comment (jorisvr) [inline tests/tools/test_proc.py]: In the present state, I think this test suite does not add any value. To say that "we have unittests for proc.py" would be deceiving ourselves.

If we intend to extend this into a somewhat serious testsuite for the proc.py code, we should create an issue to finish that work in the future. If we don't intend to do that, this file should be removed.

File: qmi/core/config_struct.py [hunk 4]
```diff
@@ -138,29 +99,12 @@ def config_struct_to_dict(cfg: object) -> dict:
...
-            elif isinstance(v, (list, tuple)):
-                v_new = _dictify_list_value(v)
-            elif isinstance(v, dict):
-                v_new = _dictify_dict_value(v)
-            elif (v is None) or isinstance(v, (int, float, str, bool)):
-                v_new = v
-            else:
-                raise TypeError("Unsupported value in field {}: {!r}".format(f.name, v))
-            ret[f.name] = v_new
+    ret = dataclasses.asdict(cfg)
```
[the remark this answers is not in the document]
Comment (qraa) [reply qmi/core/config_struct.py]: I reverted back to original implementation but rewritten the recursive function a bit to clear things up.

File: qmi/tools/proc.py [hunk 5]
```diff
@@ -124,85 +127,93 @@ def __init__(self, host: str) -> None:
...
-            self._proc.kill()
+            self._proccess.kill()
 
-        self._proc = None
+        self._proccess = None
+
+    def _proc(self):
+        if self._proccess is None:
+            raise ProcessException('Replace with nice exception')        
+        return self._proccess
```
[the remark this answers is not in the document]
Comment (qraa) [reply qmi/tools/proc.py]: removed the possibility to be None instead

File: qmi/tools/proc.py (hunk 1 above)
Comment (qraa) [reply qmi/tools/proc.py]: reverted back to original solution plus ignore types and created ticket

File: tests/tools/test_proc.py (hunk 3 above)
Comment (qraa) [reply tests/tools/test_proc.py]: created ticket

File: qmi/instruments/Tektronix_FCA3000.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ def __init__(self, context: QMI_Context, name: str, transport: str) -> None:
 
     @rpc_method
     def open(self) -> None:
+        assert self._scpi is not None
```
[the remark this answers is not in the document]
Comment (jorisvr) [reply qmi/instruments/Tektronix_FCA3000.py]: looks like this is still not right

File: qmi/tools/proc.py [hunk 7]
```diff
@@ -124,6 +128,9 @@ def __init__(self, host: str) -> None:
     def close(self) -> None:
         """Close the remote server process."""
 
+        if self._proc is None:
+            raise ProcessException("Try to close an already closed process.")
+
```
Comment (jorisvr) [inline qmi/tools/proc.py]: But we now never set `self._proc` to None, not even in close(). So this if statement will never trigger.

**Pull request merged**